### ISSUE TITLE
Documentation doesn't show how to customize built-in elements + attachShadow = bug on customized built-in elements

### ISSUE THREAD

**Issue by denis-migdal** [brython-dev/brython#2234]
Hi,

Brython documentation seems to be lacking in lot of ways.

In JS it is possible to customize built-in elements, however no example are given by the Brython documentation :
```python
class MyTR(html.TR):
	def __init__(self):

		#self.shadow = self.attachShadow({'mode': 'open'})
		self <= html.TD() # add some stuff

webcomponent.define("my-tr", MyTR)
```
```html
<table>
	<tr is="my-tr"></tr>
</table>
```

Brython documentation only shows an autonomous custom element with a shadow root.
However using a shadow root in a customize built-in elements leads to the following errors (cf below) while in JS it seems to cause no issues :
```html
<script type="text/javascript">
class MyTR extends HTMLTableRowElement {

	constructor() {
		super();

		this.attachShadow({ mode: "open" });
	}

}

	customElements.define('my-tr', MyTR, {is: "tr"});
</script>
```

```
Javascript error
 DOMException: Operation is not supported
    func http://localhost:3000/Brython-3.11.2/brython.js:13721
    __init__10 http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:62
    MyTR http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function line 82 > eval:14
    define http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:135
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    read12 http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:129
    <anonymous> http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:196
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:338
    $factory http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:319
    _request_without_body http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:357
    get http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:452
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    anonymous http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:175
    loop http://localhost:3000/Brython-3.11.2/brython.js:4762
    run_script http://localhost:3000/Brython-3.11.2/brython.js:4599
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js:4731
    ajax_load_script http://localhost:3000/Brython-3.11.2/brython.js:4729
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4714
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4705
    idb_open http://localhost:3000/Brython-3.11.2/brython.js:4690
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    brython http://localhost:3000/Brython-3.11.2/brython.js:4583
    onload http://localhost:3000/:1
```
```
DOMNode.__getattribute__/func@http://localhost:3000/Brython-3.11.2/brython.js:13721:14
__init__10@http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:62:129
MyTR@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function line 82 > eval:14:31
define@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:135:20
$B.$call1/<@http://localhost:3000/Brython-3.11.2/brython.js:5201:35
read12@http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:129:96
@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:196:25
anonymous/ajax.$factory/xmlhttp.onreadystatechange@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:338:18
EventHandlerNonNull*anonymous/ajax.$factory@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:319:5
_request_without_body@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:357:21
get@http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:452:27
$B.$call1/<@http://localhost:3000/Brython-3.11.2/brython.js:5201:35
anonymous@http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:175:82
$B.loop@http://localhost:3000/Brython-3.11.2/brython.js:4762:53
$B.run_script@http://localhost:3000/Brython-3.11.2/brython.js:4599:17
$B.loop@http://localhost:3000/Brython-3.11.2/brython.js:4776:10
$B.ajax_load_script/req.onreadystatechange@http://localhost:3000/Brython-3.11.2/brython.js:4731:1
EventHandlerNonNull*$B.ajax_load_script@http://localhost:3000/Brython-3.11.2/brython.js:4729:1
$B.loop@http://localhost:3000/Brython-3.11.2/brython.js:4776:10
$B.idb_open/idb_cx.onsuccess/openCursor.onsuccess@http://localhost:3000/Brython-3.11.2/brython.js:4714:1
EventHandlerNonNull*$B.idb_open/idb_cx.onsuccess@http://localhost:3000/Brython-3.11.2/brython.js:4705:1
EventHandlerNonNull*$B.idb_open@http://localhost:3000/Brython-3.11.2/brython.js:4690:1
$B.loop@http://localhost:3000/Brython-3.11.2/brython.js:4776:10
$B.parser.brython@http://localhost:3000/Brython-3.11.2/brython.js:4583:35
onload@http://localhost:3000/:1:1
```
```
frames 
Array [ (5) […], (5) […] ]
​
0: Array(5) [ "read", {…}, "__main__", … ]
​
1: Array(5) [ "__init__", {…}, "__main__", … ]
​
length: 2
​
<prototype>: Array []
```
```
NotSupportedError: Operation is not supported brython.js:14392:56
    flush http://localhost:3000/Brython-3.11.2/brython.js:14392
    method http://localhost:3000/Brython-3.11.2/brython.js:5509
    show_error http://localhost:3000/Brython-3.11.2/brython.js:8042
    handle_error http://localhost:3000/Brython-3.11.2/brython.js:8046
    MyTR http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function line 82 > eval:29
    define http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:135
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    read12 http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:129
    <anonyme> http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:196
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:338
    (Asynchrone : EventHandlerNonNull)
    $factory http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:319
    _request_without_body http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:357
    get http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:452
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    anonymous http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:175
    loop http://localhost:3000/Brython-3.11.2/brython.js:4762
    run_script http://localhost:3000/Brython-3.11.2/brython.js:4599
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js:4731
    (Asynchrone : EventHandlerNonNull)
    ajax_load_script http://localhost:3000/Brython-3.11.2/brython.js:4729
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4714
    (Asynchrone : EventHandlerNonNull)
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4705
    (Asynchrone : EventHandlerNonNull)
    idb_open http://localhost:3000/Brython-3.11.2/brython.js:4690
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    brython http://localhost:3000/Brython-3.11.2/brython.js:4583
    onload http://localhost:3000/:1
```
```
Uncaught DOMException: Operation is not supported
    func http://localhost:3000/Brython-3.11.2/brython.js:13721
    __init__10 http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:62
    MyTR http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function line 82 > eval:14
    define http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:135
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    read12 http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:129
    <anonymous> http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:196
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:338
    $factory http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:319
    _request_without_body http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:357
    get http://localhost:3000/Brython-3.11.2/brython.js line 9460 > Function:452
    call1 http://localhost:3000/Brython-3.11.2/brython.js:5201
    anonymous http://localhost:3000/Brython-3.11.2/brython.js line 4762 > Function:175
    loop http://localhost:3000/Brython-3.11.2/brython.js:4762
    run_script http://localhost:3000/Brython-3.11.2/brython.js:4599
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onreadystatechange http://localhost:3000/Brython-3.11.2/brython.js:4731
    ajax_load_script http://localhost:3000/Brython-3.11.2/brython.js:4729
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4714
    onsuccess http://localhost:3000/Brython-3.11.2/brython.js:4705
    idb_open http://localhost:3000/Brython-3.11.2/brython.js:4690
    loop http://localhost:3000/Brython-3.11.2/brython.js:4776
    brython http://localhost:3000/Brython-3.11.2/brython.js:4583
    onload http://localhost:3000/:1
brython.js:13721
```

Cordially,

**Comment by PierreQuentel** [comment]
Bonjour Denis :-)

Thanks for reporting this. I confess I didn't know about the "is" attribute and the 3rd attribute of define()...

I tried your example
```xml
<!doctype html>
<html>
<head>
<meta charset="utf-8">
</head>

<body>

<table>
	<tr is="my-tr"></tr>
</table>

<script>
class MyTR extends HTMLTableRowElement {

	constructor() {
		super();

		this.attachShadow({ mode: "open" });
	}

}

customElements.define('my-tr', MyTR, {is: "tr"});
</script>

</body>
</html>
```

and got this error message
```
Uncaught TypeError: Illegal constructor.
    MyTR http://localhost:8000/tests/issue_2234_js.html:17
    <anonymous> http://localhost:8000/tests/issue_2234_js.html:24
```

The [documentation](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) of `define()` seems to specify that the `options` attribute has key `extends`, not `is`, but with this change I get a different error message

```
Uncaught DOMException: Operation is not supported
    MyTR http://localhost:8000/tests/issue_2234_js.html:19
    <anonymous> http://localhost:8000/tests/issue_2234_js.html:24
```

There must be something wrong with applying `define()` to some html elements such as `<tr>`. With the `<p>` tag there is no error message

```
<!doctype html>
<html>
<head>
<meta charset="utf-8">
</head>

<body>

<p is="my-p"></p>

<script>
class MyP extends HTMLParagraphElement {

	constructor() {
		super();

		const shadow = this.attachShadow({ mode: "open" });
                shadow.appendChild(document.createTextNode('coucou'))
	}

}

customElements.define('my-p', MyP, {extends: "p"});
</script>

</body>
</html>
```

In the commit referenced above I have added the 3rd argument to `define()`; if the component class inherits a class in module browser.html, this argument is automatically added. The example below works with this commit:

```xml
<p is="my-p"></p>

<script type="text/python">
from browser import html, webcomponent

class MyP(html.P):
    def __init__(self):
        self.shadow = self.attachShadow({'mode': 'open'})
        self.shadow <= html.B('coucou') # add some stuff

webcomponent.define("my-p", MyP)

</script>
```

Can you check ?

**Comment by denis-migdal** [comment]
Bonjour ;)

> The [documentation](https://developer.mozilla.org/en-US/docs/Web/API/CustomElementRegistry/define) of `define()` seems to specify that the `options` attribute has key `extends`, not `is`,
Indeed, my bad. I got confused when I wrote the issue.

> but with this change I get a different error message
> ```
> Uncaught DOMException: Operation is not supported
>     MyTR http://localhost:8000/tests/issue_2234_js.html:19
>     <anonymous> http://localhost:8000/tests/issue_2234_js.html:24
> ```
Hummm... on Chromium I have the following error for this code :
```
test.html:19 Uncaught DOMException: Failed to execute 'attachShadow' on 'Element': This element does not support attachShadow
```

Personally, I am used to not use `attachShadow()` and instead insert the element directly. I'll check if I find more explanation on the doc. I'll keep you updated, and I'll look at the end of your message next.

**Comment by denis-migdal** [comment]
> Can you check ?
It works on Chromium and Firefox :)

> In the commit referenced above I have added the 3rd argument to `define()`; if the component class inherits a class in module browser.html, this argument is automatically added.
Great :)


For now, I used to make all my standalone custom elements class inherit `html.DIV`, in order to be able to use `this <=`, as I do not use `attachShadow()`.

Could it be possible to have like an `html.Element`  that we can inherit when doing standalone custom elements ?
This would not add the `extends` option (and the JS class would just inherit HTMLElement). Could also enable to factorize some code in all the `html.*` classes (or maybe you already have some kind of base class that could be used) ?

**Comment by denis-migdal** [comment]
I just thought of it, but `createElement()` also has another parameter ("is") :
https://developer.mozilla.org/en-US/docs/Web/API/Document/createElement

So you might need to also modify `maketag()` ?

**Comment by PierreQuentel** [comment]
> Could it be possible to have like an html.ELEMENT that we can inherit when doing standalone custom elements ?

You can use the class `browser.DOMNode`

```xml
<script type="text/python">
from browser import html, webcomponent, DOMNode

class MyP(DOMNode):
    def __init__(self):
        self <= html.B('coucou') # add some stuff

webcomponent.define("my-p", MyP)
</script>

<my-p></my-p>
```

**Comment by denis-migdal** [comment]
> You can use the class `browser.DOMNode`
Thanks, it seems perfect. ^^


For this current issue, everything is fine for me. However I think that :
- adding some more examples to Brython documentations might be useful,
- you might need to modify `maketag()` ?

I'm not closing this issue yet, as I prefer letting you close it when you think everything is fine.


I'm closing #2239 as is it is now solved.
I think #2236 is due to #2240, but maybe I should start a discussion about Template on the Google groups as having the viewpoints of others could be useful ?

In two weeks students will play with Brython, so I'll come back if they find any issues ;).


Here my project if anyone wants to play with it :
[Projet Web.zip](https://github.com/brython-dev/brython/files/12708480/Projet.Web.zip)

**Comment by denis-migdal** [comment]
Hi again ;)

Found a bug here :
```for(var base of cls.__bases__){
            if(base.__module__ == 'browser.html'){
                _extends = base.__name__.toLowerCase()
                break
            }
        }```

This will only work if the class is a direct descendant of `html.*`.
I think you need it to be recursive (something like that) :
```js
function findBrowserHTMLBase(base) {
     let stack = [base];
     while(stack.length) {
           base = stack.pop();
           if(base.__module__ == 'browser.html')
                return base.__name__.toLowerCase()
           stack.push(...base..__bases__)
     }
}

_extends = findBrowserHTMLBase(cls)
```

> you might need to modify maketag() ?
I tested `document.createElement()` in Brython, and it seems to work quite well.

Cordially

**Comment by denis-migdal** [comment]
**For reference :** I found out browsers do not like when `createElement()` is called for customized elements that insert children to themselves in their constructor (`attachShadow()` is OK though -but can't be used for some customized built-in elements e.g. when extending TR -). The recommended way is then (when `attackShadow()` isn't or can't be used)  to add the children in the `connectedCallback()` method (/!\ can be called several times).

>  The specification recommends that, as far as possible, developers should implement custom element setup in this callback rather than the constructor.

https://developer.mozilla.org/en-US/docs/Web/API/Web_components/Using_custom_elements

cf also for more explanations and examples : https://copyprogramming.com/howto/attaching-shadow-dom-to-a-custom-element-removes-error-but-why

Gotta love non-intuitive browser strange rules xD.
Maybe could be added to the doc too as it can be quite unexpected ?

**Comment by denis-migdal** [comment]
> Found a bug here :
> ...
> This will only work if the class is a direct descendant of `html.*`. I think it needs to be recursive (something like that) :


Fixed in #2295

Once PR merged, I'd close this issue and just put a link in the documentation list.

**Comment by denis-migdal** [comment]
I'm closing it as it is now merged.
